File: app/src/test/java/com/papi/nova/LayoutInflationTest.java
package com.papi.nova;

import android.content.Context;
import android.view.LayoutInflater;

import androidx.test.core.app.ApplicationProvider;

import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

@Config(sdk = {33}, shadows = {
    com.papi.nova.shadows.ShadowSpaceParticleView.class,
    com.papi.nova.shadows.ShadowBackdropFrameRenderer.class
})
@RunWith(RobolectricTestRunner.class)
public class LayoutInflationTest {
    @BeforeClass
    public static void suppressInvalidIdLogs() {
        TestLogSuppressor.install();
    }

    @Test
    public void allLayoutsInflateSuccessfully() throws IllegalAccessException {
        Context base = ApplicationProvider.getApplicationContext();
        Context context = new androidx.appcompat.view.ContextThemeWrapper(base,
                com.google.android.material.R.style.Theme_MaterialComponents_NoActionBar);
        for (int layoutId : getAllLayoutResourceIds()) {
            try {
                LayoutInflater.from(context).inflate(layoutId, null);
            } catch (android.view.InflateException e) {
                // Retry with a dummy FrameLayout for <merge> root layouts
                android.widget.FrameLayout dummyRoot = new android.widget.FrameLayout(context);
                LayoutInflater.from(context).inflate(layoutId, dummyRoot, true);
            }
        }
    }

    private static int[] getAllLayoutResourceIds() throws IllegalAccessException {
        Field[] fields = com.papi.nova.R.layout.class.getFields();
        int[] ids = new int[fields.length];
        int idx = 0;
        for (Field f : fields) {
            if (Modifier.isStatic(f.getModifiers()) && f.getType() == int.class) {
                ids[idx++] = f.getInt(null);
            }
        }
        // Trim array if necessary
        if (idx != ids.length) {
            int[] trimmed = new int[idx];
            System.arraycopy(ids, 0, trimmed, 0, idx);
            return trimmed;
        }
        return ids;
    }
}

File: app/src/test/java/com/papi/nova/LayoutInflationTest.kt
package com.papi.nova

import android.content.Context
import android.view.InflateException
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.appcompat.view.ContextThemeWrapper
import androidx.test.core.app.ApplicationProvider
import java.lang.reflect.Modifier
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@Config(
    sdk = [33],
    shadows = [
        com.papi.nova.shadows.ShadowSpaceParticleView::class,
        com.papi.nova.shadows.ShadowBackdropFrameRenderer::class
    ]
)
@RunWith(RobolectricTestRunner::class)
class LayoutInflationTest {
    @Test
    fun allLayoutsInflateSuccessfully() {
        val base = ApplicationProvider.getApplicationContext<Context>()
        val context = ContextThemeWrapper(base, com.google.android.material.R.style.Theme_MaterialComponents_NoActionBar)
        for (layoutId in getAllLayoutResourceIds()) {
            try {
                LayoutInflater.from(context).inflate(layoutId, null)
            } catch (e: InflateException) {
                val dummyRoot = FrameLayout(context)
                LayoutInflater.from(context).inflate(layoutId, dummyRoot, true)
            }
        }
    }

    companion object {
        @JvmStatic
        @BeforeClass
        fun suppressInvalidIdLogs() {
            TestLogSuppressor.install()
        }

        private fun getAllLayoutResourceIds(): IntArray {
            return R.layout::class.java.fields
                .filter { field -> Modifier.isStatic(field.modifiers) && field.type == Int::class.javaPrimitiveType }
                .map { field -> field.getInt(null) }
                .toIntArray()
        }
    }
}
